File: libs/java/src/main/java/com/mbreissi/ggcommons/messaging/MessageBuilder.java
package com.mbreissi.ggcommons.messaging;

import com.mbreissi.ggcommons.config.ConfigManager;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;

/**
 * Builder for creating Message instances with fluent API.
 *
 * <p>{@link #build()} is the single UNS identity stamping site: an explicit
 * {@link #withIdentity(MessageIdentity)} override wins; otherwise, when a config service is
 * present, the component's resolved identity ({@code ConfigManager.getComponentIdentity()}) is
 * stamped with the per-message instance token ({@link #withInstance(String)}, default
 * {@value MessageIdentity#DEFAULT_INSTANCE}); with neither, {@code identity} stays {@code null}
 * (bootstrap/raw messages legally omit it).
 */
public class MessageBuilder {
    private String name;
    private String version;
    private String correlationId;
    private Object payload;
    private ConfigManager configService;
    private String instance;
    private MessageIdentity identityOverride;

    private MessageBuilder(String name, String version) {
        this.name = name;
        this.version = version;
    }

    public static MessageBuilder create(String name, String version) {
        return new MessageBuilder(name, version);
    }

    public static Message fromObject(Object msgContents) {
        Message retVal = new Message();
        if (msgContents instanceof JsonObject msgJsonObj)
        {
            if (msgJsonObj.has("header"))
            {
                retVal.header = MessageHeader.fromDict(msgJsonObj.getAsJsonObject("header"));
            }
            if (msgJsonObj.has("identity"))
            {
                retVal.identity = Message.parseIdentity(msgJsonObj.get("identity"));
            }
            if (msgJsonObj.has("tags"))
            {
                retVal.tags = MessageTags.fromDict(msgJsonObj.getAsJsonObject("tags"));
            }
            if (msgJsonObj.has("body"))
            {
                retVal.body = msgJsonObj.get("body");
            }
            if (!(msgJsonObj.has("header") || msgJsonObj.has("identity")
                    || msgJsonObj.has("tags") || msgJsonObj.has("body")))
            {
                retVal.raw = msgJsonObj;
            }
        }
        else
        {
            retVal.raw = msgContents;
        }
        return retVal;
    }



    public MessageBuilder withCorrelationId(String correlationId) {
        this.correlationId = correlationId;
        return this;
    }

    public MessageBuilder withPayload(Object payload) {
        this.payload = payload;
        return this;
    }

    public MessageBuilder withConfig(ConfigManager configService) {
        this.configService = configService;
        return this;
    }

    /**
     * Sets the per-message instance token stamped into the identity element (default
     * {@value MessageIdentity#DEFAULT_INSTANCE}). Only takes effect when an identity is stamped
     * (a config service is present or an explicit identity override carries it).
     *
     * @param instance the instance token (non-null, non-empty)
     * @throws IllegalArgumentException if {@code instance} is null or empty
     */
    public MessageBuilder withInstance(String instance) {
        if (instance == null || instance.isEmpty()) {
            throw new IllegalArgumentException("instance must be non-empty");
        }
        this.instance = instance;
        return this;
    }

    /**
     * Sets an explicit identity override (tests, conformance vectors, relays). Wins over the
     * config-resolved identity and is stamped verbatim (the {@link #withInstance(String)} token
     * is not applied to an override).
     *
     * @param identity the identity to stamp
     */
    public MessageBuilder withIdentity(MessageIdentity identity) {
        this.identityOverride = identity;
        return this;
    }

    public Message build() {
        Message message = new Message();
        MessageHeaderBuilder headerBuilder = MessageHeaderBuilder.create(name, version);
        if (correlationId != null) {
            headerBuilder.withCorrelationId(correlationId);
        }
        message.header = headerBuilder.build();
        if (configService != null) {
            message.tags = MessageTags.fromConfig(configService);
        }

        // The single identity stamping site: explicit override > config-resolved component
        // identity (+ per-message instance token) > none (bootstrap/raw cases stay valid).
        if (identityOverride != null) {
            message.identity = identityOverride;
        } else if (configService != null) {
            MessageIdentity componentIdentity = configService.getComponentIdentity();
            if (componentIdentity != null) {
                message.identity = componentIdentity.withInstance(
                        instance != null ? instance : MessageIdentity.DEFAULT_INSTANCE);
            }
        }

        if (payload instanceof String payloadStr) {
            try {
                Gson gson = new Gson();
                message.body = gson.fromJson(payloadStr, Object.class);
            } catch (JsonSyntaxException e) {
                message.body = payloadStr;
            }
        } else {
            message.body = payload;
        }

        return message;
    }
}

File: libs/python/ggcommons/messaging/message_builder.py
"""
Builder for creating Message instances with fluent API.

``build()`` is the single UNS identity stamping site (UNS-CANONICAL-DESIGN §1.4): an
explicit :meth:`MessageBuilder.with_identity` override wins; otherwise, when a config
service is present, the component's resolved identity
(``ConfigManager.get_component_identity()``) is stamped with the per-message instance
token (:meth:`MessageBuilder.with_instance`, default ``"main"``); with neither,
``identity`` stays ``None`` (bootstrap/raw messages legally omit it).
"""
import json
from typing import TYPE_CHECKING, Optional

from ggcommons.messaging.identity import MessageIdentity

if TYPE_CHECKING:
    from ggcommons.config.manager.config_manager import ConfigManager


class MessageBuilder:
    def __init__(self, name: str, version: str):
        self.name = name
        self.version = version
        self.correlation_id = None
        self.uuid = None
        self.timestamp = None
        self.reply_to = None
        self.payload = None
        self.config_service = None
        self.tags = None
        self.instance = None
        self.identity_override: Optional[MessageIdentity] = None

    @staticmethod
    def create(name: str, version: str) -> 'MessageBuilder':
        return MessageBuilder(name, version)

    @staticmethod
    def from_object(msg_contents) -> 'MessageBuilder':
        if isinstance(msg_contents, dict) and "header" in msg_contents:
            header = msg_contents["header"]
            name = header.get("name", "unknown")
            version = header.get("version", "1.0")

            builder = MessageBuilder(name, version)

            if "correlation_id" in header:
                builder.with_correlation_id(header["correlation_id"])
            if "uuid" in header:
                builder.with_uuid(header["uuid"])
            if "timestamp" in header:
                builder.with_timestamp(header["timestamp"])
            if "reply_to" in header:
                builder.with_reply_to(header["reply_to"])
            if "identity" in msg_contents:
                # Lenient wire parse: a malformed identity yields None (no override).
                builder.with_identity(MessageIdentity.from_dict(msg_contents["identity"]))
            if "body" in msg_contents:
                builder.with_payload(msg_contents["body"])
            if "tags" in msg_contents:
                builder.with_tags(msg_contents["tags"])

            return builder
        else:
            # Raw message without header structure
            builder = MessageBuilder("raw", "1.0")
            builder.with_payload(msg_contents)
            return builder

    def with_correlation_id(self, correlation_id: str) -> 'MessageBuilder':
        self.correlation_id = correlation_id
        return self

    def with_payload(self, payload) -> 'MessageBuilder':
        self.payload = payload
        return self

    def with_config(self, config_service: 'ConfigManager') -> 'MessageBuilder':
        self.config_service = config_service
        return self

    def with_tags(self, tags: dict) -> 'MessageBuilder':
        self.tags = tags
        return self

    def with_uuid(self, uuid: str) -> 'MessageBuilder':
        """Pins the header ``uuid`` instead of the generated random one — deterministic
        envelopes for tests and the cross-language ``uns-test-vectors`` golden
        envelopes (D-U13)."""
        self.uuid = uuid
        return self

    def with_timestamp(self, timestamp: str) -> 'MessageBuilder':
        """Pins the header ``timestamp`` instead of the generated "now" — deterministic
        envelopes for tests and the cross-language ``uns-test-vectors`` golden
        envelopes (D-U13)."""
        self.timestamp = timestamp
        return self

    def with_reply_to(self, reply_to: str) -> 'MessageBuilder':
        self.reply_to = reply_to
        return self

    def with_instance(self, instance: str) -> 'MessageBuilder':
        """Sets the per-message instance token stamped into the identity element
        (default ``"main"``). Only takes effect when an identity is stamped (a config
        service is present; an explicit identity override is stamped verbatim).

        :raises ValueError: if ``instance`` is ``None`` or empty
        """
        if not instance:
            raise ValueError("instance must be non-empty")
        self.instance = instance
        return self

    def with_identity(self, identity: Optional[MessageIdentity]) -> 'MessageBuilder':
        """Sets an explicit identity override (tests, conformance vectors, relays).
        Wins over the config-resolved identity and is stamped verbatim (the
        :meth:`with_instance` token is not applied to an override)."""
        self.identity_override = identity
        return self

    def build(self):
        from ggcommons.messaging.message import Message, MessageHeader, MessageTags

        message = Message()
        message.header = MessageHeader(self.name, self.version, self.correlation_id,
                                       self.timestamp, self.uuid, self.reply_to)

        if self.tags is not None:
            message.tags = MessageTags.from_dict(self.tags)
        elif self.config_service is not None:
            message.tags = MessageTags.from_config(self.config_service)

        # The single identity stamping site (§1.4): explicit override > config-resolved
        # component identity (+ per-message instance token) > none (bootstrap/raw cases
        # stay valid).
        if self.identity_override is not None:
            message.identity = self.identity_override
        elif self.config_service is not None:
            component_identity = self.config_service.get_component_identity()
            if component_identity is not None:
                message.identity = component_identity.with_instance(
                    self.instance if self.instance else MessageIdentity.DEFAULT_INSTANCE
                )

        if isinstance(self.payload, str):
            try:
                message.body = json.loads(self.payload)
            except json.JSONDecodeError:
                message.body = self.payload
        else:
            message.body = self.payload

        return message
